introduction_to_amazon_algorithms/xgboost_abalone/xgboost_managed_spot_training.ipynb

```diff
@@ -0,0 +1,315 @@
+    "                                          train_use_spot_instances=train_use_spot_instances, \n",
+    "                                          train_max_run=train_max_run, \n",
+    "                                          train_max_wait=train_max_wait,\n",
+    "                                          checkpoint_s3_uri=checkpoint_s3_uri\n",
+    "                                         );\n",
+    "train_input = sagemaker.s3_input(s3_data='s3://{}/{}/{}'.format(bucket, prefix, 'train'), content_type='libsvm')\n",
+    "estimator.fit({'train': train_input}, job_name=job_name)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Savings\n",
+    "Towards the end of the job you should see two lines of output printed:\n",
+    "\n",
+    "- `Training seconds: X` : This is the actual compute-time your training job spent\n",
+    "- `Billable seconds: Y` : This is the time you will be billed for after Spot discounting is applied.\n",
+    "\n",
+    "If you enabled the `train_use_spot_instances`, then you should see a notable difference between `X` and `Y` signifying the cost savings you will get for having chosen Managed Spot Training. This should be reflected in an additional line:\n",
+    "- `Managed Spot Training savings: (1-Y/X)*100 %`"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Enabling checkpointing for script mode\n",
+    "\n",
+    "An additional mode of operation is to run customizable scripts as part of the training and inference jobs. See [this notebook](./xgboost_abalone_dist_script_mode.ipynb) for details on how to setup script mode. \n",
+    "\n",
+    "Here we highlight the specific changes that would enable checkpointing and use Spot instances. \n",
+    "\n",
+    "Checkpointing in the framework mode for SageMaker XGBoost can be performed using two convenient functions: \n",
+    "\n",
+    "- `save_checkpoint`: this returns a callback function that performs checkpointing of the model for each round. This is passed to XGBoost as part of the [`callbacks`](https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.train) argument. \n",
+    "\n",
+    "- `load_checkpoint`: This is used to load existing checkpoints to ensure training resumes from where it previously stopped. \n",
+    "\n",
+    "Both functions take the checkpoint directory as input, which in the below example is set to `/opt/ml/checkpoints`. \n",
+    "The primary arguments that change for the `xgb.train` call are \n",
+    "\n",
+    "1. `xgb_model`: This refers to the previous checkpoint (saved from a previously run partial job) obtained by `load_checkpoint`. This would be `None` if no previous checkpoint is available. \n",
+    "2. `callbacks`: This contains a function that performs the checkpointing\n",
+    "\n",
+    "Updated script looks like the following. \n",
+    "\n",
+    "---------\n",
+    "```\n",
```

Comment:
We've previously discussed making an `add_checkpointing()` function for script mode, and I just realized the change has been sitting in my branch and I didn't actually make a PR for this. Not sure why I completely forgot about this, my apologies.

PR here: https://github.com/aws/sagemaker-xgboost-container/pull/47

The above PR would change this code block to look like as follows:
```python
from sagemaker_xgboost_container.callback import add_checkpointing
CHECKPOINTS_DIR = '/opt/ml/checkpoints'   # default location for Checkpoints
callbacks = []
prev_checkpoint, n_remaining_iterations = add_checkpointing(
    callbacks, CHECKPOINTS_DIR, num_boost_round)
    params,
    dtrain, 
    num_boost_round=n_remaining_iterations,  
    xgb_model=prev_checkpoint,  
    callbacks=callbacks  
    )  
```